### PR TITLE
 CASSANDRA-18725: IsolatedJmx should only remove the endpoints it created when stopping (3.11)

### PR DESCRIPTION
**Pull request by JeetKunDoug**
CASSANDRA-18725: IsolatedJmx should only remove the endpoints it created when stopping

    IsolatedJmx previously cleared the entire TCPEndpoint map field, which could remove the endpoints from other,
    running instances. Now, we make sure to only clear the endpoints created by that instance, which allows for
    other instances to keep running and for this instance to restart without issues.
    Also reworked the JMX tests to exercise more of the JMX functionality to help find leaks and other
    potential issues like this.

    Additional changes include setting up `looperEverythingTest` to add increased testing like the JMX looper
    test did, so we can add more extensive exercise of native/gossip/network features in the future.

### CODE REVIEW

**Review by jonmeredith — APPROVED** [review]

File: src/java/org/apache/cassandra/utils/RMIClientSocketFactoryImpl.java [hunk 1]
```diff
@@ -42,7 +46,21 @@ public RMIClientSocketFactoryImpl(InetAddress localAddress)
...
     {
-        return new Socket(localAddress, port);
+        Socket socket = new Socket(localAddress, port);
+        sockets.add(socket);
+        return socket;
+    }
+
+    public void close() throws IOException
+    {
+        for (Socket socket: sockets) {
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/utils/RMIClientSocketFactoryImpl.java]: @JeetKunDoug please put braces on new lines in this method

File: src/java/org/apache/cassandra/utils/ReflectionUtils.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                // but for debugging purposes it is much easier to keep it like this.
+                Iterator<Map.Entry<K,V>> it = map.entrySet().iterator();
+                while (it.hasNext()) {
+                    Map.Entry<K,V> entry = it.next();
+                    if (shouldRemove.test(entry)) {
+                        it.remove();
+                    }
+                }
+            }
+        } catch (NoSuchFieldException | IllegalAccessException ex) {
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/utils/ReflectionUtils.java]: @JeetKunDoug brace

File: test/distributed/org/apache/cassandra/distributed/impl/IsolatedJmx.java [hunk 3]
```diff
@@ -127,45 +131,26 @@ public void startJmx()
...
-            }
-            catch (Throwable thrown)
-            {
-                lastThrown = thrown;
-            }
-            inInstancelogger.info("Could not connect to JMX on {} after {} attempts. Will retry.", url, attempts);
-            Thread.sleep(1000);
+        try (JMXConnector ignored = JMXUtil.getJmxConnector(config, 20, env)) {
+            // Do nothing - JMXUtil now retries
+        } catch (IOException iex) {
```
Comment (smiklosovic) [inline test/distributed/org/apache/cassandra/distributed/impl/IsolatedJmx.java]: @JeetKunDoug here too

File: test/distributed/org/apache/cassandra/distributed/test/jmx/JMXGetterCheckTest.java [hunk 4]
```diff
@@ -29,44 +29,67 @@
...
+     * Useful for more than just testing getters, it also is used in JMXFeatureTest
+     * to make sure we've touched the complete JMX code path.
+     * @param cluster the cluster to test
+     * @throws Exception several kinds of exceptions can be thrown, mostly from JMX infrastructure issues.
+     */
+    public static void testAllValidGetters(Cluster cluster) throws Exception
+    {
+        for (IInvokableInstance instance: cluster)
+        {
+            if (instance.isShutdown()) {
```
Comment (smiklosovic) [inline test/distributed/org/apache/cassandra/distributed/test/jmx/JMXGetterCheckTest.java]: @JeetKunDoug this actually does not need to be enclosed with `{ }`.

File: src/java/org/apache/cassandra/utils/ReflectionUtils.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            // noinspection unchecked
+            Map<K, V> map = (Map<K, V>) mapField.get(instance);
+            // Because multiple instances can be shutting down at once,
+            // synchronize on the map to avoid ConcurrentModificationException
+            synchronized (map)
+            {
+                // This could be done with a simple `map.entrySet.removeIf()` call
+                // but for debugging purposes it is much easier to keep it like this.
+                Iterator<Map.Entry<K,V>> it = map.entrySet().iterator();
+                while (it.hasNext()) {
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/utils/ReflectionUtils.java]: here too, `it.remove()` does not need to be put in `{ }` as it is the only statement in that `if`.

Comment (JeetKunDoug) [reply src/java/org/apache/cassandra/utils/ReflectionUtils.java]: WRT the braces where they aren't needed, the code style states they _may_ be elided not that they _must_ be - by all means, I'll fix the placement (although obviously my personal habit is to have them on the end of the line not a new line) but I think keeping single-line bodies of if statements in braces is accepted by the code style and (for me at least) helps with readability, as it's clear there's a block of code there and not just some bad indentation.

**Pull request closed without merging**